PurchasesCoreSwift/Purchasing/PurchasesOrchestrator.swift

```diff
@@ -0,0 +1,506 @@
+            let nsError = error as NSError
+            let userCancelled = nsError.code == SKError.paymentCancelled.rawValue
+            operationDispatcher.dispatchOnMainThread {
+                completion(transaction,
+                           nil,
+                           ErrorUtils.purchasesError(withSKError: error),
+                           userCancelled)
+            }
+        }
+
+        if finishTransactions {
+            storeKitWrapper.finishTransaction(transaction)
+        }
+    }
+
+    func handleDeferredTransaction(_ transaction: SKPaymentTransaction) {
+        let userCancelled: Bool
+        if let error = transaction.error as NSError? {
+            userCancelled = error.code == SKError.paymentCancelled.rawValue
+        } else {
+            userCancelled = false
+        }
+
+        guard let completion = getAndRemovePurchaseCompletedCallback(forTransaction: transaction) else {
+            return
+        }
+
+        operationDispatcher.dispatchOnMainThread {
+            completion(transaction, nil, ErrorUtils.paymentDeferredError(), userCancelled)
+        }
+    }
+
+}
+
+// MARK: Private funcs.
+private extension PurchasesOrchestrator {
+
+    func getAndRemovePurchaseCompletedCallback(forTransaction transaction: SKPaymentTransaction) -> PurchaseCompletedBlock? {
+        if let productIdentifier = transaction.rc_productIdentifier {
+            let maybeCompletion = purchaseCompleteCallbacksByProductID.removeValue(forKey: productIdentifier)
+            return maybeCompletion
+        }
+        return nil
+    }
+
+    func fetchProductsAndPostReceipt(withTransaction transaction: SKPaymentTransaction, receiptData: Data) {
+        guard let productIdentifier = transaction.rc_productIdentifier else {
+            self.handleReceiptPost(withTransaction: transaction,
+                                   maybePurchaserInfo: nil,
+                                   maybeSubscriberAttributes: nil,
+                                   maybeError: ErrorUtils.unknownError())
+            return
+        }
+
+        self.products(withIdentifiers: [productIdentifier]) { products in
+            self.postReceipt(withTransaction: transaction,
+                             receiptData: receiptData,
+                             products: Set(products))
+        }
+    }
+
+    func postReceipt(withTransaction transaction: SKPaymentTransaction, receiptData: Data, products: Set<SKProduct>) {
+        var maybeProductInfo: ProductInfo?
+        var maybePresentedOfferingID: String?
+        if let product = products.first {
+            let productInfo = ProductInfoExtractor().extractInfo(from: product)
+            let productID = productInfo.productIdentifier
+            let presentedOfferingID = presentedOfferingIDsByProductID[productID]
+            presentedOfferingIDsByProductID.removeValue(forKey: productID)
+            maybeProductInfo = productInfo
```

Comment:
can we just immediately set `ProductInfoExtractor().extractInfo(from:product)` to maybeProductInfo? and same with maybePresentedOfferingID?

Reply:
If we do that, then we'll need to handle optionality again. The reason being, we use the non-optional value from `ProductInfoExtractor().extractInfo(from: product)` for various things in that scope. 
I did, however, change where `maybeProductInfo` and `maybePresentedOfferingID` get set so they are right at the point where we have a value for them.

Reply:
wasn't reading it correctly, makes sense!